### PR TITLE
feat: add bedrock claude chat

### PR DESCRIPTION
**Pull request by bwmatson**
Replace this comment with:
  - Description: Add Bedrock implementation of Anthropic Claude for Chat
  - Tag maintainer: @hwchase17, @baskaryan
  - Twitter handle: @bwmatson

### CODE REVIEW

**Review by baskaryan — COMMENTED** [review]
this is awesome! is anthropic only chat model served by Bedrock? if not (or if that won't always be the case) could it make sense to make a slightly more generic BedrockChat, a la the Bedrock LLM class?

cc @3coins

**Comment by bwmatson** [comment]
@baskaryan I added a BedrockChatBase class to make adding more chat models to Bedrock easier.

File: langchain/chat_models/anthropic.py
Comment (hwchase17) [inline langchain/chat_models/anthropic.py:22]: this probably also needs to be moved

**Review by hwchase17 — COMMENTED** [review]
this seems good to me modulo one nit - i can make the change

**Review by hwchase17 — APPROVED** [review]

**Comment by 3coins** [comment]
@bwmatson 
Thanks for working on this. We are planning to add a generic BedrockChat class under chat_models; someone from AWS is already working on this. I notice that you have tried to capture some properties from the Anthropic class, I would like to use composition than inheritance to solve this. Can you hold off on merging this until next week?

**Comment by hwchase17** [comment]
> @bwmatson Thanks for working on this. We are planning to add a generic BedrockChat class under chat_models; someone from AWS is already working on this. I noticed that you have tried to capture some properties from the Anthropic class, I would like to use composition than inheritance to solve this. Can you hold off on merging this until next week?

sure this - will defer to you @3coins

**Review by hwchase17 — DISMISSED** [review]
wiating for larger changes from @3coins

**Comment by bwmatson** [comment]
> @bwmatson Thanks for working on this. We are planning to add a generic BedrockChat class under chat_models; someone from AWS is already working on this. I noticed that you have tried to capture some properties from the Anthropic class, I would like to use composition than inheritance to solve this. Can you hold off on merging this until next week?

@3coins This sounds good to me. Could you tag me on the AWS BedrockChat PR? I would like to see the approach you have in mind.

**Comment by 3coins** [comment]
Thanks all for helping with this.

**Comment by bwmatson** [comment]
@3coins It has been a while and I have not seen the AWS BedrockChat PR. I made some updates to my PR that I think is an improved approach. Perhaps you could provide some feedback on what you would like to see changed in this PR instead of creating a new one?

**Comment by 3coins** [comment]
@bwmatson 
Apologize for the radio silence; I agree, let’s go ahead with what you have so far. We can iterate on the base class in future. Please let me know if you need help testing this.

File: libs/langchain/langchain/chat_models/bedrock_anthropic.py
Comment (shabie) [inline libs/langchain/langchain/chat_models/bedrock_anthropic.py:14]: If we are using the string `amazon_bedrock` for our normal LLM class, shouldn't we use the same convention of using underscore as the separator and naming so like `amazon_bedrock_anthropic_chat`?

File: libs/langchain/langchain/chat_models/anthropic.py
Comment (shabie) [inline libs/langchain/langchain/chat_models/anthropic.py:20]: What do you guys think about making this class into a simple function (see below). For me at least it just reads a bit easier and a class name called "Converter" may be is indicative that it should just be a function?

```python
def convert_messages_to_prompt_anthropic(
    messages: List[BaseMessage],
    human_prompt: str = "\n\nHuman:",
    ai_prompt: str = "\n\nAssistant:"
) -> str:
   """Format the prompts in the expected format for the Anthropic model"""
   
    formatted_messages = []
    for message in messages:
        if isinstance(message, ChatMessage):
            formatted_messages.append(f"\n\n{message.role.capitalize()}: {message.content}")
        elif isinstance(message, HumanMessage):
            formatted_messages.append(f"{human_prompt} {message.content}")
        elif isinstance(message, AIMessage):
            formatted_messages.append(f"{ai_prompt} {message.content}")
        elif isinstance(message, SystemMessage):
            formatted_messages.append(f"{human_prompt} <admin>{message.content}</admin>")
        else:
            raise ValueError(f"Got unknown type {message}")
    
    if not isinstance(messages[-1], AIMessage):
        formatted_messages.append(f"{ai_prompt} ")
    
    formatted_prompt = "".join(formatted_messages).rstrip()
    return formatted_prompt
```

File: libs/langchain/langchain/chat_models/bedrock_anthropic.py
Comment (bwmatson) [reply libs/langchain/langchain/chat_models/bedrock_anthropic.py]: I agree with being consistent for the amazon bedrock naming and will include this.

Note: There a decent amount of llm and chat classes that have "_llm_type" properties with dashes in their name. Should these be consistent across all llm and chat classes?

File: libs/langchain/langchain/chat_models/anthropic.py
Comment (bwmatson) [reply libs/langchain/langchain/chat_models/anthropic.py]: I like this approach and refactored the code with a similar implementation.

**Comment by bwmatson** [comment]
@3coins I ended up switching to a generic BedrockChat class following similar patterns as the Bedrock class. Could you take a look at the updated implementation and provide some feedback?

File: libs/langchain/langchain/chat_models/bedrock_anthropic.py
Comment (shabie) [reply libs/langchain/langchain/chat_models/bedrock_anthropic.py]: I think so. Why does Bedrock have underscores?

**Review by shabie — APPROVED** [review]

**Comment by shabie** [comment]
The changes look good to me. I wanted to create a follow up PR to have merge issues resolved but messed up some stuff so I'll may be make another one if deemed necessary.

File: libs/langchain/langchain/chat_models/bedrock_anthropic.py
Comment (bwmatson) [reply libs/langchain/langchain/chat_models/bedrock_anthropic.py]: Not sure... It might not have been intentional.

**Comment by 3coins** [comment]
@bwmatson 
Looks great! Thanks for making all the updates.

@baskaryan 
This will add Chat model to the Bedrock LLM, and seems to work great. It has been open for a while, and @bwmatson just seemed to have rebased. Is there anything else needed to merge this?

**Comment by 3coins** [comment]
@hwchase17 @baskaryan 
@bwmatson has done a great job adding these updates. Can you please help merge this change, this is a pre-requisite to some other PRs on Bedrock integration. Thanks for your help.

**Comment by baskaryan** [comment]
looks great, thank you @bwmatson and @3coins!! for discoverability i think it could also help to add a section at the bottom of the ChatAntropic integration notebook showing how you can access via Bedrock, but that can be done in a follow-up PR

**Comment by 3coins** [comment]
> looks great, thank you @bwmatson and @3coins!! for discoverability i think it could also help to add a section at the bottom of the ChatAntropic integration notebook showing how you can access via Bedrock, but that can be done in a follow-up PR

Good idea, I will take care of this next week.

**Comment by mlucool** [comment]
Will this work correctly without https://github.com/langchain-ai/langchain/pull/9410?

**Comment by bwmatson** [comment]
> Will this work correctly without #9410?
I just tested the code after it was merged into master and everything is working as expedted.

**Pull request merged**
